### PR TITLE
Improving transcode dataset docs

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Description
Solves [#3676](https://github.com/kedro-org/kedro/issues/3676)

## Development notes

- Renamed the section in the docs, so it contains `transcoding dataset` reference;
- Provided some details on how the transcoding mechanism works;
- Kept an example with `pyspark` as recommended one;
- Added a new section with incorrect examples of using transcoding and the corresponding explanations.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by AhdraMeraliQB** [comment]
[See built docs](https://kedro--3833.org.readthedocs.build/en/3833/data/data_catalog_yaml_examples.html#transcode-dataset-read-the-same-file-using-two-different-datasets)

File: docs/source/data/data_catalog_yaml_examples.md [hunk 1]
```diff
@@ -365,7 +365,7 @@ airplanes:
 
 In this example, the default `csv` configuration is inserted into `airplanes` and then the `load_args` block is overridden. Normally, that would replace the whole dictionary. In order to extend `load_args`, the defaults for that block are then re-inserted.
 
-## Read the same file using two different datasets
+## Transcode dataset: Read the same file using two different datasets
```
Comment (AhdraMeraliQB) [inline docs/source/data/data_catalog_yaml_examples.md:368]:
```suggestion
## Transcoding datasets: Read the same file using different datasets
```

File: docs/source/data/data_catalog_yaml_examples.md [hunk 2]
```diff
@@ -398,6 +398,58 @@ In this example, Kedro understands that `my_dataframe` is the same dataset in it
 In the pipeline, Kedro uses the `spark.SparkDataset` implementation for saving and `pandas.ParquetDataset`
 for loading, so the first node outputs a `pyspark.sql.DataFrame`, while the second node receives a `pandas.Dataframe`.
```
Comment (AhdraMeraliQB) [inline docs/source/data/data_catalog_yaml_examples.md:399]: Unable to comment on the specific line above as it's unchanged, but in the catalog example only one entry has a `file_format` value - is this intentional?

**Review by AhdraMeraliQB — COMMENTED** [review]
Just gave this a preliminary run-through, will take a closer look at the wording shortly but would also like to get @noklam's thoughts as the author of the original issue

File: docs/source/data/data_catalog_yaml_examples.md (hunk 1 above)
Comment (ElenaKhaustova) [reply docs/source/data/data_catalog_yaml_examples.md:368]: I left `Transcode dataset` as it was in the previous docs version, though I agree with the suggestion. Will update.

File: docs/source/data/data_catalog_yaml_examples.md (hunk 2 above)
Comment (ElenaKhaustova) [reply docs/source/data/data_catalog_yaml_examples.md:399]: I think `pandas.ParquetDataset` assumes only `parquet` format: https://github.com/kedro-org/kedro-plugins/blob/0847aecb999683dada6909712e787f6de1fc2414/kedro-datasets/kedro_datasets/pandas/parquet_dataset.py#L80

File: docs/source/data/data_catalog_yaml_examples.md [hunk 3]
```diff
@@ -365,11 +365,15 @@ airplanes:
 
 In this example, the default `csv` configuration is inserted into `airplanes` and then the `load_args` block is overridden. Normally, that would replace the whole dictionary. In order to extend `load_args`, the defaults for that block are then re-inserted.
 
-## Read the same file using two different datasets
+## Read the same file using different datasets using transcoding
```
Comment (noklam) [inline docs/source/data/data_catalog_yaml_examples.md:368]:
```suggestion
## Read the same file using different datasets with transcoding
```

File: docs/source/data/data_catalog_yaml_examples.md [hunk 4]
```diff
@@ -365,11 +365,15 @@ airplanes:
...
 
-You might come across a situation where you would like to read the same file using two different dataset implementations (known as transcoding). For example, Parquet files can not only be loaded via the `ParquetDataset` using `pandas`, but also directly by `SparkDataset`. This conversion is typical when coordinating a `Spark` to `pandas` workflow.
+You might come across a situation where you would like to read the same file using two different `Dataset` implementations. You can achieve this by using transcoding to define separate `DataCalatlog` entries that point to the same `filepath`.
 
-Define two `DataCatalog` entries for the same dataset in a common format (for example, Parquet, JSON, CSV) in your `conf/base/catalog.yml`:
+### How to use transcoding
+
+Consider an example Parquet file. Parquet files can be loaded with both the `ParquetDataset` using `pandas`, and the `SparkDataset` directly. This conversion is typical when coordinating a `Spark` to `pandas` workflow.
```
Comment (noklam) [inline docs/source/data/data_catalog_yaml_examples.md:374]:
```suggestion
Consider an example with Parquet files. Parquet files can be loaded with both the `pandas.ParquetDataset`, and the `SparkDataset` directly. This conversion is typical when coordinating a `Spark` to `pandas` workflow.
```

File: docs/source/data/data_catalog_yaml_examples.md [hunk 5]
```diff
@@ -398,6 +402,78 @@ In this example, Kedro understands that `my_dataframe` is the same dataset in it
 In the pipeline, Kedro uses the `spark.SparkDataset` implementation for saving and `pandas.ParquetDataset`
 for loading, so the first node outputs a `pyspark.sql.DataFrame`, while the second node receives a `pandas.Dataframe`.
 
+### How *not* to use transcoding
+
+Kedro pipelines automatically resolve the node execution order and check to ensure there are no circular dependencies in the pipeline. It is during this process that the transcoded datasets are resolved and the transcoding notation `@...` is stripped. This means within the pipeline the datasets `my_dataframe@spark` and `my_dataframe@pandas` are considered to be one `my_dataframe` dataset. The `DataCatalog`, however, treats transcoded entries as separate datasets, as they are only resolved as part of the pipeline resolution process. This results in differences between your defined pipeline in `pipeline.py` and the resolved pipeline that is run by Kedro, and these differences may lead to unintended behaviours. Thus, it is important to be aware of this when using transcoding.
+
+Below are some examples where transcoding may produce unwanted side effects.
+
+#### Defining a node with the same inputs and outputs
```
Comment (noklam) [inline docs/source/data/data_catalog_yaml_examples.md:411]: Is this still possible after the change?

File: docs/source/data/data_catalog_yaml_examples.md [hunk 6]
```diff
@@ -398,6 +402,78 @@ In this example, Kedro understands that `my_dataframe` is the same dataset in it
...
+```
+
+During the pipeline resolution, the node above is defined as having the dataset `my_dataset` as both its input and output. As a node cannot have the same inputs and outputs, trying to run this pipeline will fail with the following error:
+
+```console
+ValueError: Failed to create node my_func1([my_dataframe@pandas]) -> [my_dataframe@spark].
+A node cannot have the same inputs and outputs even if they are transcoded: {'my_dataframe'}
+```
+
+#### Defining several nodes that share the same output
```
Comment (noklam) [inline docs/source/data/data_catalog_yaml_examples.md:430]: Is this possible too? I think this should not be allowed at all with pipeline validation.

**Review by noklam — COMMENTED** [review]
Leave some comments but I think we should really add validation instead of trying too hard to explain this if possible.

File: docs/source/data/data_catalog_yaml_examples.md (hunk 5 above)
Comment (ElenaKhaustova) [reply docs/source/data/data_catalog_yaml_examples.md:411]: It's not possible, and this section explains why. Below, there's also an example of an error raising.

File: docs/source/data/data_catalog_yaml_examples.md (hunk 6 above)
Comment (ElenaKhaustova) [reply docs/source/data/data_catalog_yaml_examples.md:430]: No, not possible, same as above.

**Comment by ElenaKhaustova** [comment]
> Leave some comments but I think we should really add validation instead of trying too hard to explain this if possible.

Applied suggested changes, thank you!

As for validation, it's already there. The section `How *not* to use transcoding` aims to clarify the logic behind transcoding and point to potential issues. Though it's good to know if the explanations are unclear.

File: docs/source/data/data_catalog_yaml_examples.md (hunk 5 above)
Comment (datajoely) [reply docs/source/data/data_catalog_yaml_examples.md:411]: I would put and admonition like this to make it clear to someone skim reading this quickly

> [!CAUTION]
> Put some content here .....

File: docs/source/data/data_catalog_yaml_examples.md [hunk 7]
```diff
@@ -365,11 +365,15 @@ airplanes:
 
 In this example, the default `csv` configuration is inserted into `airplanes` and then the `load_args` block is overridden. Normally, that would replace the whole dictionary. In order to extend `load_args`, the defaults for that block are then re-inserted.
 
-## Read the same file using two different datasets
+## Read the same file using different datasets with transcoding
 
-You might come across a situation where you would like to read the same file using two different dataset implementations (known as transcoding). For example, Parquet files can not only be loaded via the `ParquetDataset` using `pandas`, but also directly by `SparkDataset`. This conversion is typical when coordinating a `Spark` to `pandas` workflow.
+You might come across a situation where you would like to read the same file using two different `Dataset` implementations. You can achieve this by using transcoding to define separate `DataCalatlog` entries that point to the same `filepath`.
```
Comment (yuryfedotov-mck) [inline docs/source/data/data_catalog_yaml_examples.md:370]:
```suggestion
You might come across a situation where you would like to read the same file using two different `Dataset` implementations. You can achieve this by using transcoding to define separate `DataCatalog` entries that point to the same `filepath`.
```

**Review by datajoely — APPROVED** [review]

**Review by noklam — APPROVED** [review]

**Review by yuryfedotov-mck — APPROVED** [review]

**Review by ankatiyar — APPROVED** [review]

**Pull request merged**